relay-general/src/pii/attachments.rs

```diff
@@ -110,6 +114,259 @@ fn apply_regex_to_bytes(
+    ///
+    /// If the replacement string encodes to a longer byte-slice than the current string the
+    /// replacement string is truncated.  If this does not align with a character boundary
+    /// in the replacement string it is further trucated to the previous character boundary
+    /// and the remainder is filled with the padding char.
+    ///
+    /// If the combination of the possibly truncated replacement string plus the padding
+    /// character can not match the byte-slice of the current string `Err` is returned.
+    fn swap_content(&mut self, replacement: &str, padding: char) -> Result<(), Self::Error>;
+}
+
+impl StringMods for WStr {
+    type Error = Utf16Error;
+
+    fn fill_content(&mut self, fill_char: char) -> Result<(), Self::Error> {
+        let size = std::mem::size_of::<u16>();
+
+        if (fill_char.len_utf16() == 2) && ((self.len() / size) % 2 != 0) {
+            return Err(Utf16Error::new());
+        }
+
+        let mut buf = [0u16; 2];
+        let fill_u16 = fill_char.encode_utf16(&mut buf[..]);
+        let mut fill_bytes: Vec<u8> = Vec::with_capacity(fill_u16.len() * size);
+        for code in fill_u16 {
+            fill_bytes.extend_from_slice(&code.to_le_bytes());
+        }
+
+        debug_assert!(
+            self.len() % fill_bytes.len() == 0,
+            "string size not multiple of fill size"
+        );
+        let chunks = self.as_bytes_mut().chunks_exact_mut(fill_bytes.len());
+        for chunk in chunks {
+            chunk.copy_from_slice(fill_bytes.as_slice());
+        }
+
+        Ok(())
+    }
+
+    fn swap_content(&mut self, replacement: &str, padding: char) -> Result<(), Self::Error> {
+        // WARNING: This implementation writes to the current string before it knows things
+        // will work out.  That is incorrect!
+        let size = std::mem::size_of::<u16>();
+        let len = self.len();
+        let mut offset = 0;
+        for code in replacement.encode_utf16() {
+            let char_len = if 0xD800 & code == 0xD800 {
+                size * 2 // leading surrogate
+            } else {
+                size
+            };
+            if (len - offset) < char_len {
+                break; // Not enough space for this char
+            }
+            let target = &mut self.as_bytes_mut()[offset..offset + size];
+            target.copy_from_slice(&code.to_le_bytes());
+            offset += size;
+        }
+
+        // Our current WStr could be invalid UTF-16LE at this point, up to the current
+        // offset it is valid, but afterwards is may not be.  So we can not just slice
+        // ourselves and must slice the raw bytes, creating a new WStr with them.  This new
+        // WStr could be invalid, but after we called .fill_content() it will be valid.
+        let remainder_bytes = &mut self.as_bytes_mut()[offset..];
+        let remainder = unsafe { WStr::from_utf16le_unchecked_mut(remainder_bytes) };
+        remainder.fill_content(padding)
```

Comment:
you're specifying that constructing wstr with invalid utf16 is UB but you're also assuming that `fill_content` is able to deal with a wstr that is partially invalid, a state that should not exist. There is not much in the way of somebody modifying either `from_utf16le_unchecked_mut` or `fill_content` to segfault or panic on invalid input each.

I think that a method called like `from_*_unchecked` is typically used to bypass expensive validation of input that is already valid, not to construct types in broken states. I would not construct a `String` with broken input either, even if I just call the most trivial method on it.

Perhaps you want `fill_content` to be a static method that takes bytes by value and returns wstr?

Reply:
good catch, thanks! simplified and avoided using UB